execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,280 @@
+	roundControl             lru.BasicLRU[uint64, *expressLaneControl]
+	messagesBySequenceNumber map[uint64]*timeboost.ExpressLaneSubmission
+}
+
+func newExpressLaneService(
+	auctionContractAddr common.Address,
+	sequencerClient *ethclient.Client,
+	bc *core.BlockChain,
+) (*expressLaneService, error) {
+	chainConfig := bc.Config()
+	auctionContract, err := express_lane_auctiongen.NewExpressLaneAuction(auctionContractAddr, sequencerClient)
+	if err != nil {
+		return nil, err
+	}
+	roundTimingInfo, err := auctionContract.RoundTimingInfo(&bind.CallOpts{})
+	if err != nil {
+		return nil, err
+	}
+	initialTimestamp := time.Unix(int64(roundTimingInfo.OffsetTimestamp), 0)
+	roundDuration := time.Duration(roundTimingInfo.RoundDurationSeconds) * time.Second
+	auctionClosingDuration := time.Duration(roundTimingInfo.AuctionClosingSeconds) * time.Second
+	return &expressLaneService{
+		auctionContract:          auctionContract,
+		chainConfig:              chainConfig,
+		initialTimestamp:         initialTimestamp,
+		auctionClosing:           auctionClosingDuration,
+		roundControl:             lru.NewBasicLRU[uint64, *expressLaneControl](8), // Keep 8 rounds cached.
+		auctionContractAddr:      auctionContractAddr,
+		roundDuration:            roundDuration,
+		seqClient:                sequencerClient,
+		logs:                     make(chan []*types.Log, 10_000),
+		messagesBySequenceNumber: make(map[uint64]*timeboost.ExpressLaneSubmission),
+	}, nil
+}
+
+func (es *expressLaneService) Start(ctxIn context.Context) {
+	es.StopWaiter.Start(ctxIn, es)
+
+	// Log every new express lane auction round.
+	es.LaunchThread(func(ctx context.Context) {
+		log.Info("Watching for new express lane rounds")
+		now := time.Now()
+		waitTime := es.roundDuration - time.Duration(now.Second())*time.Second - time.Duration(now.Nanosecond())
+		time.Sleep(waitTime)
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case t := <-ticker.C:
+				round := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
+				log.Info(
+					"New express lane auction round",
+					"round", round,
+					"timestamp", t,
+				)
+				es.Lock()
+				// Reset the sequence numbers map for the new round.
+				es.messagesBySequenceNumber = make(map[uint64]*timeboost.ExpressLaneSubmission)
+				es.Unlock()
+			}
+		}
+	})
+	es.LaunchThread(func(ctx context.Context) {
+		log.Info("Monitoring express lane auction contract")
+		// Monitor for auction resolutions from the auction manager smart contract
+		// and set the express lane controller for the upcoming round accordingly.
+		latestBlock, err := es.seqClient.HeaderByNumber(ctx, nil)
+		if err != nil {
+			// TODO: Should not be a crit.
+			log.Crit("Could not get latest header", "err", err)
+		}
+		fromBlock := latestBlock.Number.Uint64()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(time.Millisecond * 250):
```

Comment:
I wonder if `time.After(time.Millisecond * 250)` is sufficient and whether it’s better to stream new blocks directly from the tx feed. Otherwise, there could be a race condition where, at 249.9ms, we check and find that the latest block number hasn’t changed, forcing us to wait another 250ms.